<h2>Alexander Lerma</h2>

<h2>Homework 2 - CS 422 </h2>

<h1>2 Practicum Problems</h1>
<h2>2.1 Problem 1</h2>

Load the market-basket sample dataset into the Orange application, and run both frequent itemset as well as association rule modules. Set the support threshold to 10% and observe the antecedent in the rules with the highest lift. What item is observed to be there, and what is its support? Is this a valuable association rule? Why or why not?

<i>Association Rules with Support Threshold = 10%</i>
![title](img/problem1.png)

<b>Result:</b>

We observe that Antecedant Eggs=1 with Consequent Bread=1, Beer=1,

{Eggs} -> {Bread, Beer}, had the highest lift value of 2.50

We know that having a lift > 1 corresponds to how dependent the antecedent and consequent are on each other. 

Therefore having a lift value of 2.50 indicates a high dependency on one another and is a very valuable association rule. Stores can use this data to place items in close proximity to other items that customers tend to buy together. 



<h2>2.2 Problem 2</h2>

Load the Extended Bakery dataset (75000-out2-final.csv) into the Orange application, and run both frequent itemset as well as association rule modules. Set the support threshold to 1% and the confidence threshold to 90%. Observe the association rules containing the Cherry Tart item within the antecedent. What other item appears with it? When the confidence threshold is lowered to 45%, does the Cherry Tart item now appear without another item in the antecedent? Is the same consequent observed in both cases? How did lowering the confidence threshold lead to this change? Hint: Reference the Simpson’s Paradox section of the text.

<b>Result:</b>

<i>Association Rules with Support Threshold = 10%, Confidence = 90%</i>
![title](img/problem2a.png)

<i>Association Rules with Support Threshold = 10%, Confidence = 45%</i>
![title](img/problem2b.png)



With support=10% and confidence=90%, we observe that only Opera Cake is in the Antecedent with Cherry Tart and yields only one association rule involving Cherry Tart in the Antecedent:

{Opera Cake, Cherry Tart} -> {Appricot Danish}

With support=10% and confidence=45%, we observe that Cherry Tart now appears without another item in the Antecedent producing the rules:

{Cherry Tart} -> {Opera Cake}

{Cherry Tart} -> {Apricot Danish}

An additional consequent, Opera Cake, is now observed when confidence=45%

A lower confidence threshold of 45% can be generalized as follows. If we were to repeat our procedure for finding associations on multiple samples of data, the <i>calculated</i> cofidence interval, which varies for each sample and is not the same as the <i>overall</i> confidence interval, will contain the given association rule 45% of the time. If an association occurs that lies outside the calculated confidence interval, then there was <= 65% chance of that association being made. Lowering the confidence interval allowed an association rule to be revealed that wouldn't make the cut when working with a 90% confidence interval. This relates closely with the Simpson's Paradox.

<h2>2.3 Problem 3 </h2>

Load the Extended Bakery dataset (75000-out2-binary.csv) into Python using a Pandas dataframe. Calculate the binary correlation coefficient Φ for the Chocolate Coffee and Chocolate Cake items. Show whether the two items are symmetric binary variables via their co-presence and co-absence. Would an as- sociation rule between these items as antecedent and consequent have a high confidence level? Why or why not?

In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import matthews_corrcoef
%matplotlib inline

In [18]:
df = pd.read_csv('dataset/75000-out2-binary.csv')
df.head(5)

Transaction Number  Chocolate Cake  Lemon Cake  Casino Cake  Opera Cake  \
0                   1               0           0            0           0   
1                   2               0           0            0           0   
2                   3               0           0            0           1   
3                   4               0           0            0           0   
4                   5               0           0            0           0   

   Strawberry Cake  Truffle Cake  Chocolate Eclair  Coffee Eclair  \
0                0             0                 0              0   
1                0             0                 0              1   
2                0             0                 0              0   
3                0             1                 0              0   
4                0             0                 1              0   

   Vanilla Eclair       ...         Lemon Lemonade  Raspberry Lemonade  \
0               0       ...                      0                   0   
1               0       ...                      0                   0   
2               0       ...                      0                   0   
3               0       ...                      0                   0   
4               0       ...                      0                   0   

   Orange Juice  Green Tea  Bottled Water  Hot Coffee  Chocolate Coffee  \
0             0          0              0           0                 0   
1             0          0              0           1                 0   
2             1          0              0           0                 0   
3             0          0              0           0                 0   
4             1          0              0           0                 0   

   Vanilla Frappuccino  Cherry Soda  Single Espresso  
0                    0            0                0  
1                    0            0                0  
2                    0            0                0  
3                    1            0                0  
4                    0            0                0  

[5 rows x 51 columns]

In [19]:
df.describe()

Transaction Number  Chocolate Cake    Lemon Cake   Casino Cake  \
count        75000.000000    75000.000000  75000.000000  75000.000000   
mean         37500.500000        0.083533      0.083613      0.075013   
std          21650.779432        0.276689      0.276809      0.263415   
min              1.000000        0.000000      0.000000      0.000000   
25%          18750.750000        0.000000      0.000000      0.000000   
50%          37500.500000        0.000000      0.000000      0.000000   
75%          56250.250000        0.000000      0.000000      0.000000   
max          75000.000000        1.000000      1.000000      1.000000   

         Opera Cake  Strawberry Cake  Truffle Cake  Chocolate Eclair  \
count  75000.000000     75000.000000  75000.000000      75000.000000   
mean       0.082093         0.092640      0.082240          0.042373   
std        0.274509         0.289929      0.274732          0.201441   
min        0.000000         0.000000      0.000000          0.000000   
25%        0.000000         0.000000      0.000000          0.000000   
50%        0.000000         0.000000      0.000000          0.000000   
75%        0.000000         0.000000      0.000000          0.000000   
max        1.000000         1.000000      1.000000          1.000000   

       Coffee Eclair  Vanilla Eclair       ...         Lemon Lemonade  \
count   75000.000000    75000.000000       ...           75000.000000   
mean        0.109240        0.042520       ...               0.068240   
std         0.311942        0.201774       ...               0.252159   
min         0.000000        0.000000       ...               0.000000   
25%         0.000000        0.000000       ...               0.000000   
50%         0.000000        0.000000       ...               0.000000   
75%         0.000000        0.000000       ...               0.000000   
max         1.000000        1.000000       ...               1.000000   

       Raspberry Lemonade  Orange Juice     Green Tea  Bottled Water  \
count        75000.000000  75000.000000  75000.000000   75000.000000   
mean             0.067747      0.091613      0.062467       0.075200   
std              0.251312      0.288481      0.242003       0.263716   
min              0.000000      0.000000      0.000000       0.000000   
25%              0.000000      0.000000      0.000000       0.000000   
50%              0.000000      0.000000      0.000000       0.000000   
75%              0.000000      0.000000      0.000000       0.000000   
max              1.000000      1.000000      1.000000       1.000000   

         Hot Coffee  Chocolate Coffee  Vanilla Frappuccino   Cherry Soda  \
count  75000.000000      75000.000000         75000.000000  75000.000000   
mean       0.102667          0.083147             0.077467      0.061987   
std        0.303525          0.276106             0.267332      0.241133   
min        0.000000          0.000000             0.000000      0.000000   
25%        0.000000          0.000000             0.000000      0.000000   
50%        0.000000          0.000000             0.000000      0.000000   
75%        0.000000          0.000000             0.000000      0.000000   
max        1.000000          1.000000             1.000000      1.000000   

       Single Espresso  
count     75000.000000  
mean          0.067973  
std           0.251702  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           0.000000  
max           1.000000  

[8 rows x 51 columns]

In [20]:
# creates a confusion matrix
def confusion_matrix(x, y):
    matrix = np.array([[0, 0],
                       [0, 0]])
    for i in range(min(len(x), len(y))):
        matrix[x[i]][y[i]] += 1
    return matrix

# computes binary coefficient Phi
# Note: I also computed this using sklearn which produced a slightly different result
def compute_phi_coefficient(x, y):
    matrix = confusion_matrix(x, y)
    a = (matrix[1][1] * matrix[0][0]) - (matrix[1][0] * matrix[0][1])
    b = np.sqrt(sum(matrix[1]) * sum(matrix[0]) * sum(matrix[:, 0]) * sum(matrix[:, 1]))
    return np.divide(a, b)

In [21]:
X = df['Chocolate Coffee']
Y = df['Chocolate Cake']

# using sklearn's 'Mathew's Correlation Coefficient' (same as phi)
matthews_corrcoef(X, Y)

0.48556649252787693

In [22]:
# using my function for computing phi
compute_phi_coefficient(X, Y)

0.48556649252787681

In [23]:
# copresence 
cm = confusion_matrix(X, Y)
total = min(len(X), len(Y))
copresence = cm[1][1] 
coabsence = cm[0][0]
diagonal = copresence + coabsence
print('Copresence ratio: {} / {} or {}%'.format(copresence, total, np.divide(copresence, total) * 100.))
print('Coabsence ratio: {} / {} or {}%'.format(coabsence, total, np.divide(coabsence, total) * 100.))
print('Total ratio falling along diagonal cells: {} / {} or {}%'.format(diagonal, total, np.divide(diagonal, total) * 100.))

Copresence ratio: 3303 / 75000 or 4.404%
Coabsence ratio: 65802 / 75000 or 87.736%
Total ratio falling along diagonal cells: 69105 / 75000 or 92.14%


<b>Result</b>

A generaliztion of phi coefficient on Wikipedia
(https://en.wikipedia.org/wiki/Phi_coefficient):

"Two binary variables are considered positively associated if most of the data falls along the diagonal cells."

We observed that 92.14% of the data was either copresent or coabsent. 

conf(Chocolate Cake implies Chocolate Coffee) = 3303 / 75000 or 4.404%

conf(not Chocolate Cake implies not Chocolate Coffee) = 65802 / 75000 or 87.736%

So we can create the association rule {not Chocolate Cake} -> {not Chocolate Coffee} from the 87.736% calculated confidence, but would not want to create a rule for the copresent due it having only a 4.404% calculated confidence.
